api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -180,7 +180,32 @@ message PrivateKeyProvider {
   }
 }
 
-// [#next-free-field: 9]
+// [#not-implemented-hide:]
+// BoringSSL private key method configuration. The private key methods are used for external
+// (potentially asynchronous) signing and decryption operations. Some use cases for private key
+// methods would be TPM support and TLS acceleration.
+message PrivateKeyProviderEntry {
+  // Private key method provider name. The name must match a
+  // supported private key method provider type.
+  string provider_name = 1 [(validate.rules).string = {min_len: 1}];
+}
+
+// [#not-implemented-hide:]
+// Provides a list of private key providers. Envoy will find out an available private
+// key provider from the list on order. If there is none of available private key provider,
+// it may fallback to BoringSSL default implementation based on the `fallback` fallback.
+message PrivateKeyProviderList {
+  // A list of private key providers, and at least one private key provider provided.
+  repeated PrivateKeyProviderEntry private_key_provider = 1
```

Comment:
I would have naively though this would be repeated `PrivateKeyProvider` why is that insufficient?

Reply:
My mistake, I should use the `PrivateKeyProvider` directly. I thought the private key is part of `PrivateKeyProvider` message. Actually, it is not. It is part of each provider's config.

So I should deprecate this field https://github.com/envoyproxy/envoy/blob/c49f392561ba8c246f845d61e555c9536d5edbea/api/contrib/envoy/extensions/private_key_providers/qat/v3alpha/qat.proto#L29

And I want all the providers using the private key here https://github.com/envoyproxy/envoy/blob/c49f392561ba8c246f845d61e555c9536d5edbea/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L200

I didn't see a reason why each provider needs its own private key. If we have a list of providers here, they can share the same private key.

Reply:
I added some description about which fields are going to deprecate in the commit message
https://github.com/envoyproxy/envoy/pull/28215#issue-1785483852